embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/telemetry/EmbraceTelemetryService.kt

```diff
@@ -0,0 +1,77 @@
+package io.embrace.android.embracesdk.telemetry
+
+import io.embrace.android.embracesdk.capture.metadata.EmbraceMetadataService
+import io.embrace.android.embracesdk.internal.spans.toEmbraceAttributeName
+import java.util.Locale
+
+/*
+    Service for tracking usage of public APIs, and different internal metrics about the app.
+ */
+internal class EmbraceTelemetryService(
+    private val okHttpReflectionFacade: OkHttpReflectionFacade = OkHttpReflectionFacade()
+) {
+
+    private val usageCountMap = mutableMapOf<String, Int>()
+    private val appAttributesMap = mutableMapOf<String, String>()
+
+    /*
+        Tracks the usage of a public API by name. Adds a suffix for easier identification. Replaces whitespaces with underscores.
+    */
+    fun onPublicApiCalled(name: String) {
+        val suffixedName = "emb.usage.${
+        name.toLowerCase(Locale.ENGLISH)
+            .trim()
+            .replace(" ", "_")
+        }"
+        usageCountMap[suffixedName] = (usageCountMap[suffixedName] ?: 0) + 1
+    }
+
+    /*
+        Returns a map with every telemetry value. For now, it's just usage counts of public APIs.
+     */
+    fun getTelemetryAttributes(): Map<String, String> {
+        val telemetryMap = mutableMapOf<String, String>()
+
+        telemetryMap.putAll(getUsageCountTelemetry())
+
+        telemetryMap.putAll(getAppAttributes())
+
+        return telemetryMap
+    }
+
+    private fun getUsageCountTelemetry() = usageCountMap.mapValues {
```

Comment:
I think `mapValues` ultimately creates an iterator which means this is likely to throw `ConcurrentModificationException` if other threads insert values while this is iterating. This seems likely given the scope of places this can be called. Using `ConcurrentHashMap` or similar would help avoid this

Reply:
+1

Reply:
Converting the usageCountMap to ConcurrentHashMap will be enough? 

I am now doing this for getting the attributes. Should I change anything else?

```
        usageCountMap
            .entries.associate {
                it.key.toEmbraceUsageAttributeName() to it.value.toString()
            }
```

Also, should this be a ConcurrentHashMap? Lazy already handles concurrency right?

```
private val appAttributes: Map<String, String> by lazy { getAppAttributes() }
```

Reply:
`ConcurrentHashMap` takes a collection as a constructor parameter so I'd advise creating one from the result of whatever you've got now.

`lazy` handles concurrency around initialising a variable multiple times. It doesn't alter how the underlying collection works - and by default maps aren't thread-safe in Kotlin

Reply:
usageCountMap is now a ConcurrentHashMap!

I'm still unsure about appAttributes: if lazy ensures that it will only be initialized by one thread, and after that the value will be cached, wouldn't it be already thread safe?

Reply:
Regarding `appAttributes`, since we return an immutable map, concurrent access should be fine. And since `lazy` blocks on the invocation of the supplier method, we should be good there too?

Reply:
Regarding `usageCountMap`, is the concern that access to `usageCountMap` isn't synchronized, so even if the underlying data structure is thread safe (i.e. cool to access on multiple threads), you can be in the middle of `getAndClearTelemetryAttributes()`  and have someone call `onPublicApiCalled()` and cause a `ConcurrentModificationException`?

Reply:
I think if we can make the snapshotting and clearing of `usageCountMap` atomic (and minimize the creation of new map instances), this should be fine? Do don't want to simply synchronize the two public methods because we want to preserve the concurrent invocations of `onPublicApiCalled()`, so we'll have to be a bit more thoughtful around this.